code_generator/GdkCodeGenerator/Templates/UnityComponentConversionGenerator.tt

```diff
@@ -9,493 +9,333 @@
-            translation.EntityIdTo<#= eventDetails.EventName #>Events[eventSender.EntityId].Add(eventData);
-        }
+            public ComponentReplicator(EntityManager entityManager) : base(entityManager)
+            {
+            }
+
+            public override void ExecuteReplication(ComponentGroup replicationGroup, Connection connection)
+            {
+                var componentData = replicationGroup.GetComponentDataArray<<#= componentDetails.TypeName #>>();
+                var entityIdData = replicationGroup.GetComponentDataArray<SpatialEntityId>();
+
+                for (var i = 0; i < componentData.Length; i++) 
+                {
+                    var data = componentData[i];
+                    if (data.DirtyBit)
+                    {
+                        var update = new global::Improbable.Worker.Core.SchemaComponentUpdate(<#= componentDetails.ComponentId #>);
+                        <#= componentDetails.TypeName #>.Serialization.Serialize(data, update.GetFields());
+                        connection.SendComponentUpdate(entityIdData[i].EntityId, new global::Improbable.Worker.Core.ComponentUpdate(update));
+
+                        // TODO: Generate event serialization into update
+
+                        data.DirtyBit = false;
+                        componentData[i] = data;
+                    }
+                }
+            }
+
+            public override void SendCommands(List<ComponentGroup> commandComponentGroups, Connection connection)
+            {
+                // TODO: Generate command serialization
+            }
 
-        public static List<<#= eventDetails.FullyQualifiedPayloadTypeName #>> Get<#= eventDetails.EventName #>Events(this EventSender<<#= componentDetails.TypeName #>> eventSender)
-        {
-            var translation = <#= componentDetails.ComponentName #>.Translation.GetTranslation(eventSender.InternalHandleToTranslation);
-            return translation.EntityIdTo<#= eventDetails.EventName #>Events[eventSender.EntityId];
         }
 
-        public static void Clear<#= eventDetails.EventName #>Events(this EventSender<<#= componentDetails.TypeName #>> eventSender)
+        public class ComponentCleanup : ComponentCleanupHandler 
         {
-            var translation = <#= componentDetails.ComponentName #>.Translation.GetTranslation(eventSender.InternalHandleToTranslation);
-            translation.EntityIdTo<#= eventDetails.EventName #>Events[eventSender.EntityId].Clear();
+            public override ComponentType[] CleanUpComponentTypes => cleanUpComponentTypes;
+            private static readonly ComponentType[] cleanUpComponentTypes = {
+                typeof(ComponentAdded<<#= componentDetails.TypeName #>>),
+                typeof(ComponentRemoved<<#= componentDetails.TypeName #>>)
+                // TODO: More to come potentially
```

Comment:
Give the IL2CPP problems it may be a good idea to generate lists for all the components that can be removed.

Reply:
Yes - I expect this will become concrete and not go down the reflected method path

Reply:
That was the original plan - but IL2CPP cannot resolve the reflection used for that so will probably want to sidestep that